node/src/actors/chain_manager/mining.rs

```diff
@@ -314,6 +325,47 @@ impl ChainManager {
                     }
                 })
                 .flatten()
+                // Refrain from trying to resolve any more requests if we have already hit the limit
+                // of retrievals per epoch.
+                .and_then(move |vrf_proof| {
+                    let mut local_retrieval_count = cloned_retrieval_count.load(Relaxed);
+
+                    if local_retrieval_count + additional_retrieval_count > maximum_retrieval_count {
+                        log::info!("{} Refrained from resolving data request {} for epoch #{} because it contains {} \
+                        sources, which added to the sources that have already been retrieved ({}) would total {} \
+                        retrievals, which exceed current limit per epoch ({}). This limit exists for performance and \
+                        security reasons. You can increase the limit (AT YOUR OWN RISK) by adjusting the \
+                        `data_request_max_retrievals_per_epoch` inside the `[mining]` section in the `witnet.toml` \
+                        configuration file.",
+                            Yellow.bold().paint("[Mining]"),
+                            Yellow.bold().paint(dr_pointer.to_string()),
+                            Yellow.bold().paint(current_epoch.to_string()),
+                            Yellow.bold().paint(additional_retrieval_count.to_string()),
+                            Yellow.bold().paint(local_retrieval_count.to_string()),
+                            Yellow.bold().paint((local_retrieval_count + additional_retrieval_count).to_string()),
+                            Yellow.bold().paint(maximum_retrieval_count.to_string())
+                        );
+
+                        Err(())
+                    } else {
+                        // Update `current_retrieval_count` thanks to interior mutability of the
+                        // `cloned_retrieval_count` reference. This is a recursive operation so as
+                        // to guarantee addition and prevent potential race conditions in a concurrent
+                        // scenario.
+                        loop {
+                            let internal_retrieval_count = cloned_retrieval_count.compare_and_swap(local_retrieval_count, local_retrieval_count + additional_retrieval_count, Relaxed);
+                            if internal_retrieval_count == local_retrieval_count {
+                                // The counter update was updated successfully, we can move on.
+                                break;
+                            } else {
+                                // The counter was updated somewhere else, addition must be retried.
+                                local_retrieval_count = internal_retrieval_count;
```

Comment:
Here we have to check ` if local_retrieval_count + additional_retrieval_count > maximum_retrieval_count` again